File: 2017/03.py

import math, copy, re

MAX = 1000
HALF = 500
SPIRAL = [[None for j in range(MAX)] for i in range(MAX)]
SPIRAL[500][500] = 1
dirs = 'enws'

def part_1(data):
	n = math.sqrt(data) 
	if n == n // 1:
		print(n - 1)
		return
	else:
		n = n // 1
		n += 1 if n % 2 == 0 else 2
	half = n // 2
	end = n ** 2
	start = (n - 2) ** 2 + 1
	tmp = data - start
	diff = n - 1
	half_side = (diff - 1) // 2


	a = half
	b = abs((tmp % diff) - half_side)
	manhattan = a + b
	print(manhattan)

	return

def get_neighbours_num(pos):
	[row, col] = pos
	cnt = 0
	for i in range(row - 1, row + 2):
		for j in range(col - 1, col + 2):
			if i == row and j == col: continue
			if SPIRAL[i][j] is not None: cnt += 1
	return cnt

def get_neighbours_sum(pos):
	[row, col] = pos
	suma = 0
	for i in range(row - 1, row + 2):
		for j in range(col - 1, col + 2):
			if i == row and j == col: continue
			if SPIRAL[i][j] is not None: suma += SPIRAL[i][j]
	return suma

def move(position, direction):
	[row, col] = position
	if direction == 'e':
		col += 1
	elif direction == 'w':
		col -= 1
	elif direction == 'n':
		row += 1
	elif direction == 's':
		row -= 1
	return [row, col]

def part_2(data):
	curr_dir = 'e'
	curr_pos = [HALF, HALF]
	suma = 0
	while True:
		curr_pos = move(curr_pos, curr_dir)
		if (suma := get_neighbours_sum(curr_pos)) >= data:
			print(suma)
			break
		print(suma)
		SPIRAL[curr_pos[0]][curr_pos[1]] = suma
		neig_num = get_neighbours_num(curr_pos)
		if neig_num <= 2:
			curr_dir = dirs[(dirs.index(curr_dir) + 1) % len(dirs)]
	return 


if __name__ == '__main__':
	with open('3_input') as f:
		data = f.read()
		# data = data.split('\n')
		# data = list(map(int, data.split()))
		data = int(data)

	part_1(data)
	part_2(data)
	

File: 2017/03.js

const fs = require('fs');
const { exec } = require("child_process");
const { ALPHA_L, ALPHA_U, D4, D6, D8, MOD } = require("../lib");
const { joseph, findCycle } = require("../lib");
const { cache } = require("../lib");
const { bin, float, hex, int, num, oct } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { copy,	entries, in_, inBB, keys, sort, values } = require("../lib");
const { Counter, DD, empty, FastQueue, PriorityQueue, Queue, set, Stack } = require("../lib");
const { areaInt, circumference, manDist, shoelace } = require("../lib");
const { ord, chr, count, debug, disp, disp3, crt, gcd, lcm, modPow, modPowBig, modInv, mod, prod, prodBig, randint, sum, sumBig, transpose } = require("../lib");
const { digits, ints,	floats,	singles,	words,	lines,	table,	groups,	getGroups,	groupsWith,	parse,	parseLine, scanf } = require ('../lib');
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { isSuperset, or, and, xor, sub } = set;
const { getExecStr } = require("../lib/post");
const { combinations, combinations_with_replacement, next_permutation, product } = require("../lib");

function part1(data) {

	let res = 0;
	data = int(data);
	let sq = floor(sqrt(data)) + 1;
	if (sq % 2 === 0) ++sq;
	let big = sq ** 2, small = (sq - 2) ** 2;
	let tmp = big;
	let cnt = 0;
	while (tmp > data) {
		tmp -= (sq - 1);
		++cnt;
	}
	debug((sq - 1) / 2, data, tmp)
	res = (sq - 1) / 2 + abs((sq - 1) / 2 - (data - tmp));

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART1");
	return;
}

const fillData = (G, data) => {
	const l = G.length;
	let x = int(l/2), y = int(l/2);
	let i = 0;
	let dy = 0, dx = 1;
	G[y][x] = 1;
	const left = (_G, _y, _x, _dy, _dx) => {
		const [ndy, ndx] = [dy ? 0 : -dx, dx ? 0 : dy];
		return G[y + ndy][x + ndx] ? [dy, dx] : [ndy, ndx];
	}
	while (true) {
		y += dy;
		x += dx;
		let sm = 0;
		drange(3)((i, j) => {
			sm += G[y - 1 + i][x - 1 + j];
		});
		G[y][x] = sm;
		if (sm > data) return sm;
		[dy, dx] = left(G, y, x, dy, dx);
	}
}

function part2(data) {

	let res = 0;
	data = int(data);
	let sq = floor(sqrt(data));
	if (sq % 2 === 0) ++sq;
	const GRID = new Array(sq).fill(0).map(r => new Array(sq).fill(0));
	res = fillData(GRID, data);

	debug(res);
	if (res) exec(`echo ${res} | xclip -sel clip -rmlastnl`);
	console.log("END OF PART2");
	return;
}

function main() {
	let data = fs.readFileSync("03_input").toString("utf-8");

	part1(data);
	part2(data);
	process.exit(0);
}

main();
